File: packages/core/src/core/converter.ts
import { parse } from './parser.js';
import { walk } from './walker.js';
import { render } from './renderer.js';
import { deduplicateBlocks } from './dedup.js';
import { contentHash } from './hash.js';
import { extractContent } from '../extract/index.js';
import { estimateTokens } from '../tokens/index.js';
import { getDefaultRules } from '../rules/index.js';
import type { ConvertOptions, ConvertResult, ResolvedOptions } from '../types.js';

const DEFAULTS: ResolvedOptions = {
    extract: false,
    rules: [],
    baseUrl: '',
    headingStyle: 'atx',
    bulletChar: '-',
    codeBlockStyle: 'fenced',
    fenceChar: '`',
    strongDelimiter: '**',
    emDelimiter: '*',
    linkStyle: 'inlined',
    deduplicate: false
};

/**
 * Convert an HTML string to markdown.
 *
 * @param html - Raw HTML to convert.
 * @param options - Conversion settings (heading style, extraction, custom rules, etc.).
 * @returns The generated markdown and an estimated token count.
 *
 * @example
 * ```ts
 * const { markdown, tokenEstimate } = convert("<h1>Hello</h1><p>World</p>");
 * ```
 */
export function convert(html: string, options?: ConvertOptions): ConvertResult {
    const opts: ResolvedOptions = { ...DEFAULTS, ...options };

    const document = parse(html);

    if (opts.extract) {
        const extractOpts = typeof opts.extract === 'object' ? opts.extract : undefined;
        extractContent(document, extractOpts);
    }

    // eslint-disable-next-line unicorn/no-array-sort -- toSorted unavailable in ES2022 target
    const allRules = [...opts.rules, ...getDefaultRules()].sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));

    const raw = walk(document, {
        options: opts,
        rules: allRules,
        listDepth: 0,
        insidePre: false,
        insideTable: false
    });

    let markdown = render(raw);

    if (opts.deduplicate) {
        const minLength = typeof opts.deduplicate === 'object' ? opts.deduplicate.minLength : undefined;
        markdown = deduplicateBlocks(markdown, minLength);
    }

    const tokenEstimate = opts.tokenCounter ? opts.tokenCounter(markdown) : estimateTokens(markdown);

    return { markdown, tokenEstimate, contentHash: contentHash(markdown) };
}

File: python/src/markdown_for_agents/core/converter.py
from __future__ import annotations

import time

from .._types import (
    ConvertResult,
    DeduplicateOptions,
    ExtractOptions,
    ResolvedOptions,
    Rule,
)
from ..extract import extract_content
from ..rules import get_default_rules
from ..tokens import estimate_tokens
from .dedup import deduplicate_blocks
from .frontmatter import extract_metadata, serialize_frontmatter
from .hash import content_hash
from .parser import parse
from .renderer import render
from .walker import WalkerState, walk

_DEFAULTS = ResolvedOptions()


def convert(
    html: str,
    *,
    extract: bool | ExtractOptions = False,
    rules: list[Rule] | None = None,
    base_url: str = "",
    heading_style: str = "atx",
    bullet_char: str = "-",
    code_block_style: str = "fenced",
    fence_char: str = "`",
    strong_delimiter: str = "**",
    em_delimiter: str = "*",
    link_style: str = "inlined",
    deduplicate: bool | DeduplicateOptions = False,
    frontmatter: bool | dict[str, str] = True,
    token_counter: object | None = None,
    server_timing: bool = False,
) -> ConvertResult:
    """Convert an HTML string to markdown."""
    start = time.perf_counter() if server_timing else 0.0

    opts = ResolvedOptions(
        extract=extract,
        rules=tuple(rules) if rules else (),
        base_url=base_url,
        heading_style=heading_style,  # type: ignore[arg-type]
        bullet_char=bullet_char,  # type: ignore[arg-type]
        code_block_style=code_block_style,  # type: ignore[arg-type]
        fence_char=fence_char,  # type: ignore[arg-type]
        strong_delimiter=strong_delimiter,  # type: ignore[arg-type]
        em_delimiter=em_delimiter,  # type: ignore[arg-type]
        link_style=link_style,  # type: ignore[arg-type]
        deduplicate=deduplicate,
        frontmatter=frontmatter,
        token_counter=token_counter,  # type: ignore[arg-type]
    )

    document = parse(html)

    # Extract metadata from <head> before extract strips it
    frontmatter_block = ""
    if opts.frontmatter is not False:
        extracted = extract_metadata(document)
        user_fields = opts.frontmatter if isinstance(opts.frontmatter, dict) else {}
        merged = {**extracted, **user_fields}
        frontmatter_block = serialize_frontmatter(merged)

    if opts.extract:
        extract_opts = opts.extract if isinstance(opts.extract, ExtractOptions) else None
        extract_content(document, extract_opts)

    # Sort rules: user rules + defaults, highest priority first
    all_rules = sorted(
        list(opts.rules) + get_default_rules(),
        key=lambda r: r.priority,
        reverse=True,
    )

    raw = walk(
        document,
        WalkerState(
            options=opts,
            rules=all_rules,
            list_depth=0,
            inside_pre=False,
            inside_table=False,
        ),
    )

    md = render(raw)

    if opts.deduplicate:
        min_length = opts.deduplicate.min_length if isinstance(opts.deduplicate, DeduplicateOptions) else 10
        md = deduplicate_blocks(md, min_length)

    # Prepend frontmatter after render/dedup
    md = frontmatter_block + md

    token_est = opts.token_counter(md) if opts.token_counter is not None else estimate_tokens(md)

    convert_duration = (time.perf_counter() - start) * 1000 if server_timing else None

    return ConvertResult(
        markdown=md,
        token_estimate=token_est,
        content_hash=content_hash(md),
        convert_duration=convert_duration,
    )
